phylo/acceleration.go

```diff
@@ -0,0 +1,339 @@
+	sub.DhcaOrang = (d.WhumOrang*(d.DhumOrang-b.BhumHca) + d.WchimpOrang*(d.DchimpOrang-b.BchimpHca)) / (d.WhumOrang + d.WchimpOrang)
+
+	sub.WgorOrang = d.DgorOrang
+	sub.WhcaGor = sub.DhcaGor
+	sub.WhcaOrang = sub.DhcaOrang
+}
+
+//optimizeSubtreeLeft is a helper function of BranchLengthsAlternatingLeastSquares and computes the optimal branch lengths for the left subtree at a particular iteration.
+func optimizeSubtreeLeft(sub *AccelSubTreeLeft, allowNegative bool, verbose bool, answer *AccelBranchLengths) {
+	sub.VhumHca = (sub.DhumChimp + sub.DhumHga - sub.DchimpHga) / 2.0
+	sub.VchimpHca = (sub.DhumChimp + sub.DchimpHga - sub.DhumHga) / 2.0
+	sub.VhcaHga = (sub.DhumHga + sub.DchimpHga - sub.DhumChimp) / 2.0
+	if allowNegative {
+		answer.BhumHca = sub.VhumHca
+		answer.BchimpHca = sub.VchimpHca
+		answer.BhcaHga = sub.VhcaHga
+		return
+	}
+	if sub.VhumHca < 0 && sub.VchimpHca < 0 && sub.VhcaHga < 0 {
+		if verbose {
+			log.Printf("WARNING: All branches are negative.") //TODO: Should this error out?
+		}
+		sub.VhumHca = 0
+		sub.VchimpHca = 0
+		sub.VhcaHga = 0
+	} else if sub.VhumHca < 0 {
+		sub.VhumHca = 0
+		if sub.VchimpHca < 0 {
+			sub.VchimpHca = 0
+			sub.VhcaHga = nonNegativeApproximation(sub.DhumHga, sub.DchimpHga, sub.VhumHca, sub.VchimpHca, sub.WhumHga, sub.WchimpHga)
+		} else if sub.VhcaHga < 0 {
+			sub.VhcaHga = 0
+			sub.VchimpHca = nonNegativeApproximation(sub.DhumChimp, sub.DchimpHga, sub.VhumHca, sub.VhcaHga, sub.WhumChimp, sub.WchimpHga)
+		} else {
+			sub.VhcaHga = nonNegativeApproximation(sub.DhumHga, sub.DchimpHga, sub.VhumHca, sub.VchimpHca, sub.WhumHga, sub.WchimpHga)
+			sub.VchimpHca = nonNegativeApproximation(sub.DhumChimp, sub.DchimpHga, sub.VhumHca, sub.VhcaHga, sub.WhumChimp, sub.WchimpHga)
+		}
+	} else if sub.VchimpHca < 0 {
+		sub.VchimpHca = 0
+		if sub.VhcaHga < 0 {
+			sub.VhcaHga = 0
+			sub.VhumHca = nonNegativeApproximation(sub.DhumHga, sub.DhumChimp, sub.VhcaHga, sub.VchimpHca, sub.WhumHga, sub.WhumChimp)
+		} else {
+			sub.VhumHca = nonNegativeApproximation(sub.DhumHga, sub.DhumChimp, sub.VhcaHga, sub.VchimpHca, sub.WhumHga, sub.WhumChimp)
+			sub.VhcaHga = nonNegativeApproximation(sub.DhumHga, sub.DchimpHga, sub.VhumHca, sub.VchimpHca, sub.WhumHga, sub.WchimpHga)
+		}
+	} else if sub.VhcaHga < 0 {
+		sub.VhcaHga = 0
+		sub.VhumHca = nonNegativeApproximation(sub.DhumHga, sub.DhumChimp, sub.VhcaHga, sub.VchimpHca, sub.WhumHga, sub.WhumChimp)
+		sub.VchimpHca = nonNegativeApproximation(sub.DhumChimp, sub.DchimpHga, sub.VhumHca, sub.VhcaHga, sub.WhumChimp, sub.WchimpHga)
+	}
+	answer.BhumHca = sub.VhumHca
+	answer.BchimpHca = sub.VchimpHca
+	answer.BhcaHga = sub.VhcaHga
+}
+
+//optimizeSubtreeRight is a helper function of BranchLengthsAlternatingLeastSquares and computes the optimal branch lengths for the right subtree at a particular iteration.
+func optimizeSubtreeRight(sub *AccelSubTreeRight, allowNegative bool, verbose bool, answer *AccelBranchLengths) {
```

Comment:
Can we reduce the probability of bugs by having one optimizeSubtree function that is general to a three-species subtree, or is it better to have two separate functions to do the left and right?

Reply:
that was how it was done in the old version. The trade off on bugs is having more confusing variable names passed around into the generalized function. I wanted the explicit left and right optimization function so you could see precisely what was passed in and passed back. Happy to do it either way